consensus/hotstuff/signature/randombeacon_reconstructor.go

```diff
@@ -1,52 +1,85 @@
+// execute the business logic, without interfering with each other).
+// It allows concurrent verification of the given signature.
+// Returns :
+//  - engine.InvalidInputError if signerIndex is invalid
+//  - module/signature.ErrInvalidFormat if signerID is valid but signature is cryptographically invalid
+//  - other error if there is an unexpected exception.
 func (r *RandomBeaconReconstructor) Verify(signerID flow.Identifier, sig crypto.Signature) error {
-	panic("to be implemented")
+	signerIndex, err := r.dkg.Index(signerID)
+	if err != nil {
+		return fmt.Errorf("could not map signerID %v to signerIndex: %w", signerID, err)
+	}
+	return r.randomBeaconInspector.Verify(int(signerIndex), sig)
 }
 
-// TrustedAdd adds the signature share to the reconstructors internal
-// state. Validity of signature is not checked. It is up to the
-// implementation, whether it still adds a signature or not, when the
-// minimal number of required sig shares has already been reached,
-// because the reconstructed group signature is the same.
-// Returns: true if and only if enough signature shares were collected
+// TrustedAdd adds a share to the internal signature shares store.
+// There is no pre-check of the signature's validity _before_ adding it.
+// It is the caller's responsibility to make sure the signature was previously verified.
+// Nevertheless, the implementation guarantees safety (only correct threshold signatures
+// are returned) through a post-check (verifying the threshold signature
+// _after_ reconstruction before returning it).
+// The function is thread-safe but locks its internal state, thereby permitting only
+// one routine at a time to add a signature.
+// Returns:
+//  - (true, nil) if the signature has been added, and enough shares have been collected.
+//  - (false, nil) if the signature has been added, but not enough shares were collected.
+//  - (false, error) if there is any exception adding the signature share.
+//      - engine.InvalidInputError if signerIndex is invalid
+//  	- engine.DuplicatedEntryError if the signer has been already added
+//      - other error if there is an unexpected exception.
 func (r *RandomBeaconReconstructor) TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (bool, error) {
-	panic("to be implemented")
+	signerIndex, err := r.dkg.Index(signerID)
+	if err != nil {
+		return false, fmt.Errorf("could not map signerID %v to signerIndex: %w", signerID, err)
+	}
+	return r.randomBeaconInspector.TrustedAdd(int(signerIndex), sig)
 }
 
-// EnoughShares returns true if and only if reconstructor
-// has collected a sufficient number of signature shares.
+// EnoughShares indicates whether enough shares have been accumulated in order to reconstruct
+// a group signature. The function is thread-safe.
 func (r *RandomBeaconReconstructor) EnoughShares() bool {
-	panic("to be implemented")
+	return r.randomBeaconInspector.EnoughShares()
```

Comment:
if `hotstuff.RandomBeaconInspector` was declared to be embedded inside `RandomBeaconReconstructor`, we won't need to write this function, it will be just inherited. 

```
type RandomBeaconReconstructor struct {
	dkg                   hotstuff.DKG                   
	hotstuff.RandomBeaconInspector
}
```